nextroute/schema/fleet.go

```diff
@@ -0,0 +1,234 @@
+	LatenessPenalty         *float64     `json:"lateness_penalty,omitempty"`
+	CompatibilityAttributes *[]string    `json:"compatibility_attributes,omitempty"`
+}
+
+// FleetToNextRoute takes a legacy cloud fleet input and converts it into
+// nextroute input format.
+func FleetToNextRoute(fleetInput FleetInput) (Input, error) {
+	input := Input{}
+	stopCompats := make([]string, 0)
+	vehicleCompats := make([]string, 0)
+	// Use default values and add special handling for CompatibilityAttributes
+	// and HardWindows.
+	if fleetInput.Defaults != nil && fleetInput.Defaults.Stops != nil {
+		input.Defaults = &Defaults{}
+		input.Defaults.Stops = &StopDefaults{
+			UnplannedPenalty:        fleetInput.Defaults.Stops.UnassignedPenalty,
+			Quantity:                fleetInput.Defaults.Stops.Quantity,
+			MaxWait:                 fleetInput.Defaults.Stops.MaxWait,
+			Duration:                fleetInput.Defaults.Stops.StopDuration,
+			TargetArrivalTime:       fleetInput.Defaults.Stops.TargetTime,
+			EarlyArrivalTimePenalty: fleetInput.Defaults.Stops.EarlinessPenalty,
+			LateArrivalTimePenalty:  fleetInput.Defaults.Stops.LatenessPenalty,
+		}
+		if fleetInput.Defaults.Stops.CompatibilityAttributes != nil {
+			stopCompats = *fleetInput.Defaults.Stops.CompatibilityAttributes
+		}
+		if fleetInput.Defaults.Stops.HardWindow != nil {
+			input.Defaults.Stops.StartTimeWindow = *fleetInput.Defaults.Stops.HardWindow
+		}
+	}
+
+	// Use default values. nil values are not compatible between fleet and
+	// nextroute.
+	if fleetInput.Defaults != nil && fleetInput.Defaults.Vehicles != nil {
+		vehicleCompats = fleetInput.Defaults.Vehicles.CompatibilityAttributes
+		input.Defaults.Vehicles = &VehicleDefaults{
+			Capacity:          fleetInput.Defaults.Vehicles.Capacity,
+			StartLocation:     fleetInput.Defaults.Vehicles.Start,
+			EndLocation:       fleetInput.Defaults.Vehicles.End,
+			Speed:             fleetInput.Defaults.Vehicles.Speed,
+			StartTime:         fleetInput.Defaults.Vehicles.ShiftStart,
+			EndTime:           fleetInput.Defaults.Vehicles.ShiftEnd,
+			MaxStops:          fleetInput.Defaults.Vehicles.MaxStops,
+			MaxDistance:       fleetInput.Defaults.Vehicles.MaxDistance,
+			MaxDuration:       fleetInput.Defaults.Vehicles.MaxDuration,
+			StartLevel:        nil,
+			MinStops:          nil,
+			MinStopsPenalty:   nil,
+			MaxWait:           nil,
+			ActivationPenalty: nil,
+		}
+	}
+
+	// Handle special case of compatibility attributes, to use them to make
+	// legacy backlogs more or less work like in legacy fleet.
+	for i := range fleetInput.Vehicles {
+		fleetInput.Vehicles[i].CompatibilityAttributes =
+			append(fleetInput.Vehicles[i].CompatibilityAttributes, vehicleCompats...)
+	}
+
+	// Create vehicles with special logic for backlog legacy needs.
+	backlogStops := make(map[string]struct{})
+	vehicles := make([]Vehicle, len(fleetInput.Vehicles))
+	for i, v := range fleetInput.Vehicles {
+		v := v
```

Comment:
Wut?

Reply:
this is a workaround gecause gosec complains because a pointer of v is passed and v is a loop variable. go does reuse the pointers for variable. this assignment gives each loop variable its own address.

Reply:
Why would a pointer of v be passed? I'm not seeing it I guess?

Reply:
its a pointer to a field in v. but that comes to the same result.

Reply:
Kk.